src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,90 +539,179 @@ internal override void Complete()
-                                }
-                            }
-                            if (_outputReader != null)
-                            {
-                                ProcessOutputHelper();
-                            }
-                        }
-                    }
                 }
                 catch (Exception)
                 {
                     StopProcessing();
                     throw;
                 }
-                finally
+
+                if (_isRunningInBackground == false)
+                {
+                    InitOutputQueue();
+                }
+            }
+            catch (Win32Exception e)
+            {
+                exceptionToRethrow = e;
+
+            } // try
+            catch (PipelineStoppedException)
+            {
+                // If we're stopping the process, just rethrow this exception...
+                throw;
+            }
+            catch (Exception e)
+            {
+                CommandProcessorBase.CheckForSevereException(e);
+
+                exceptionToRethrow = e;
+            }
+
+            // An exception was thrown while attempting to run the program
+            // so wrap and rethrow it here...
+            if (exceptionToRethrow != null)
+            {
+                // It's a system exception so wrap it in one of ours and re-throw.
+                string message = StringUtil.Format(ParserStrings.ProgramFailedToExecute,
+                    this.NativeCommandName, exceptionToRethrow.Message,
+                    this.Command.MyInvocation.PositionMessage);
+                ApplicationFailedException appFailedException = new ApplicationFailedException(message, exceptionToRethrow);
+
+                // There is no need to set this exception here since this exception will eventually be caught by pipeline processor.
+                // this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
+
+                throw appFailedException;
+            }
+        }
+
+        private void InitOutputQueue()
+        {
+            //if output is redirected, start reading output of process in queue.
+            if (_nativeProcess.StartInfo.RedirectStandardOutput || _nativeProcess.StartInfo.RedirectStandardError)
+            {
+                lock (_sync)
+                {
+                    if (!_stopped)
+                    {
+                        _nativeProcessOutputQueue = new BlockingCollection<ProcessOutputObject>();
+                        // we don't assign the handler to anything, because it's used only for objects marshaling
+                        new ProcessOutputHandler(_nativeProcess, _nativeProcessOutputQueue);
+                    }
+                }
+            }
+        }
+
+        private ProcessOutputObject DequeueProcessOutput(bool blocking)
+        {
+            if (blocking)
+            {
+                if (!_nativeProcessOutputQueue.IsCompleted)
```

Comment:
Can we please keep the old comment `// if adding was completed and the queue is empty, there is no need to do a blocking Take()`? It will help reader to understand why we check `IsCompleted` which is a great optimization to avoid unnecessary exception from `Take()`.